atc/worker/artifact_source_test.go

```diff
@@ -24,6 +27,78 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+var _ = Describe("ArtifactSourcer", func() {
+	var (
+		logger          *lagertest.TestLogger
+		fakeCompression *compressionfakes.FakeCompression
+	)
+
+	BeforeEach(func() {
+		logger = lagertest.NewTestLogger("test")
+		fakeCompression = new(compressionfakes.FakeCompression)
+	})
+
+	It("locates images by handle", func() {
+		artifact := runtime.GetArtifact{VolumeHandle: "image"}
+		vf := FakeVolumeFinder{Volumes: map[string]worker.Volume{
+			"image": newVolumeWithContent(content{".": []byte("image content")}),
+		}}
+
+		sourcer := worker.NewArtifactSourcer(fakeCompression, vf, false, 0)
+		source, err := sourcer.SourceImage(logger, artifact)
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(source).To(BeStreamableWithContent(content{".": []byte("image content")}))
+	})
+
+	It("locates inputs and caches", func() {
```

Comment:
i could for sure, but I wanted to experiment with breaking away from the Ginkgo/BDD `BeforeEach` pattern a little here. I'm mostly just dissatisfied with the tests I've written/that are prevalent in the codebase (in terms of how difficult it is to refactor the code under test), so wanted to try something new. I watched https://youtu.be/8hQG7QlcLBk?t=1216, and one suggestion that sounded interesting (albeit controversial) was `Repeat yourself` - it's okay to copy/paste some setup code if it makes it easier to debug the tests.

admittedly, the test setup here is a little awkward, especially faking out `FindVolumeForTaskCache`. but I'm imagining a future world where we test in terms of the high-level constructs, e.g.

```go
pool := newFakePool(
    newWorker("worker1").withVolumes(newEmptyVolume("existing_cache")),
    newWorker("worker2").withVolumes(newVolumeWithContent("output", content{".": "output content"}))
)
...
sourcer := worker.NewArtifactSourcer(compression, pool, false, 0)
inputSources, err := sourcer.SourceInputsAndCaches(logger, 0, inputs)
...
```

so that way we're actually testing the surface area of the API rather than the internal abstractions (`worker.Worker`, in this case)

---

...so, I think I'll keep it like this for now! lemme know if you disagree and I don't mind changing it, just letting you know where my head's at w.r.t. testing!

Reply:
Totally fine with that! I've had the same gripe with a lot of our test suites as well and have thought about what it would be like to follow a less dry way of doing things. (just watched the first few seconds of that vid and the second point he makes of "logic spread across many callers" has me going "YASSSS THAT SUCKS!!". The pattern of Before/JustBefore drives me crazy quite often. Jumping up and down the file is so annoying -__-

Alright, we're leaving this as-is then! This is why I approved it initially :D